cmd/shadowsocks-local/local.go

```diff
@@ -1,430 +1,814 @@
+			_, err = serverListener.WriteTo(readBuf[length:n], clientAddr)
+			if err != nil {
+				return
+			}
 		}
 	}()
+	defer serverListener.Close()
+	defer ssPacketConn.Close()
+
+	cliReq := make([]byte, UDPMaxSize)
+	for {
+		// TODO FIXME need to determine when close the TCP connection
+		// server read the first packet from associated address, get the client destination address
+		// ss-local will not support the fragments about udp, we will forward and log a WARNING
+		// if the FRAG is not 0
+		n, raddr, err := serverListener.ReadFrom(cliReq)
+		if err != nil {
+			ss.Logger.Error("[UDP] error in read packet from client UDP", zap.Error(err))
+			return
+		}
+		cliReq = cliReq[:n]
+		getAddr, err := net.ResolveUDPAddr("udp", raddr.String())
+		if getAddr.Port != clientPort {
+			ss.Logger.Warn("[UDP] read a illegal source packet, droped")
+			continue
+		}
+
+		// if frag != 0 should get a warning
+		if cliReq[idFrag]&0xff != 0x00 {
+			ss.Logger.Warn("[UDP] warrning get the fragnent packet")
+		}
 
-	go ss.PipeThenClose(conn, remote)
-	ss.PipeThenClose(remote, conn)
-	closed = true
-	debug.Println("closed connection to", addr)
+		// here we get the request need to encrypte and forward to ss remote
+		// write them into the Nat table
+		var reqAddrLen int
+		switch cliReq[idType] & 0xff {
+		case typeIPv4:
+			reqAddrLen = lenIPv4
+		case typeIPv6:
+			reqAddrLen = lenIPv6
+		case typeDm:
+			reqAddrLen = lenDmBase + int(cliReq[idType+1])
+		}
+		var rawData []byte = cliReq[idType:reqAddrLen]
+		if _, ok := NatInfo[serverPort]; !ok {
+			NatInfo[serverPort] = make(natTable)
+		}
+		NatInfo[serverPort][string(rawData)] = &natTableInfo{clientPort, serverPort, serverListener, ssPacketConn}
+
+		// write to the ss-remote, this data will be encrypted with the choosen server
+		_, err = ssPacketConn.WriteTo(cliReq[idType:], serverUDPaddr)
+		if err != nil {
+			ss.Logger.Error("[UDP]write to server error", zap.Stringer("ss-local", ssPacketConn.LocalAddr()),
+				zap.String("server", server), zap.Error(err))
+			continue
+		}
+		ss.Logger.Debug("closed server connection", zap.Stringer("ss-local", ssPacketConn.LocalAddr()), zap.String("server", server))
+	}
 }
 
-func run(listenAddr string) {
-	ln, err := net.Listen("tcp", listenAddr)
+var ciphers map[string]encrypt.Cipher
+
+func run(config *ss.Config) {
+	laddr := net.JoinHostPort(config.Local, strconv.Itoa(config.LocalPort))
+	ln, err := net.Listen("tcp", laddr)
+	if err != nil {
+		ss.Logger.Fatal("error in shadwsocks local server listen", zap.Error(err))
+	}
+	ss.Logger.Info("starting local socks5 server", zap.String("listenAddr", laddr))
+
+	// prepare connect to the server ciphers
+	cps, err := prepareToConnect(config)
 	if err != nil {
-		log.Fatal(err)
+		ss.Logger.Fatal("error in shadwsocks local server prepaer the cipher", zap.Error(err))
+	}
+
+	// init the ciphers map
+	ciphers = cps
+
+	servers := config.GetServerArray()
```

Comment:
redundant lines and manipulations

Reply:
any question here?

Reply:
config.go is much complex , cause I start the reconstruct from it, now this need to rework

Reply:
@dszhengyu fixed in 8796767